compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -291,170 +364,222 @@ class Semantic {
               Result(obj.fields(target), Nil)
             else
               value.select(target, source, needResolve = false)
 
-        case Fun(body, thisV, klass) =>
+        case Fun(body, thisV, klass, env) =>
           // meth == NoSymbol for poly functions
           if meth.name.toString == "tupled" then Result(value, Nil) // a call like `fun.tupled`
-          else eval(body, thisV, klass, cacheResult = true)
+          else
+            withEnv(env) {
+              eval(body, thisV, klass, cacheResult = true) ++ checkArgs
+            }
 
         case RefSet(refs) =>
-          val resList = refs.map(_.call(meth, superType, source))
+          val resList = refs.map(_.call(meth, args, superType, source))
           val value2 = resList.map(_.value).join
           val errors = resList.flatMap(_.errors)
           Result(value2, errors)
       }
+    }
 
     /** Handle a new expression `new p.C` where `p` is abstracted by `value` */
-    def instantiate(klass: ClassSymbol, ctor: Symbol, source: Tree): Contextual[Result] =
-      value match {
+    def instantiate(klass: ClassSymbol, ctor: Symbol, args: List[ArgInfo], source: Tree): Contextual[Result] = log("instantiating " + klass.show + ", value = " + value + ", args = " + args, printer, res => res.asInstanceOf[Result].show) {
+      val trace1 = trace.add(source)
+      if promoted.isCurrentObjectPromoted then Result(Hot, Nil)
+      else value match {
         case Hot  =>
-          Result(Hot, Errors.empty)
+          val buffer = new mutable.ArrayBuffer[Error]
+          val args2 = args.map { arg =>
+            val errors = arg.promote
+            buffer ++= errors
+            if errors.isEmpty then Hot
+            else arg.value.widenArg
+          }
+
+          if buffer.isEmpty then Result(Hot, Errors.empty)
+          else
+            val value = Warm(klass, Hot, ctor, args2).ensureExists
+            val res = value.call(ctor, args, superType = NoType, source)
+            Result(value, res.errors)
 
         case Cold =>
-          val error = CallCold(ctor, source, trace.toVector)
+          val error = CallCold(ctor, source, trace1.toVector)
           Result(Hot, error :: Nil)
 
         case addr: Addr =>
+          given Trace = trace1
+
           // widen the outer to finitize addresses
           val outer = addr match
-            case Warm(_, _: Warm) => Cold
+            case Warm(_, _: Warm, _, _) => Cold
             case _ => addr
 
-          val value = Warm(klass, outer)
-          if !heap.contains(value) then
-            val obj = Objekt(klass, fields = mutable.Map.empty, outers = mutable.Map(klass -> outer))
-            heap.update(value, obj)
-          val res = value.call(ctor, superType = NoType, source)
-          Result(value, res.errors)
+          val value = Warm(klass, outer, ctor, args.map(_.value).widenArgs).ensureExists
+          val res = value.call(ctor, args, superType = NoType, source)
```

Comment:
This is problematic: still possible to crash.

Reply:
@olhotak @EnzeXing In 41807c9 and 7f91775, I proposed two different fixes. Which fix makes more sense?

Reply:
The errors are fewer in 98a0977  because the analysis just skipped the check.

Reply:
I kind of prefer 7f91775, as the code is more regular --- we only have one special case in variable resolution. As the examples are all invented, I don't expect the two do not have an observable difference in expressiveness nor usability.

Reply:
OK, I'm OK with https://github.com/lampepfl/dotty/commit/7f917756a1abcc8b682aa2d58a38a3e8f76e72df as well. Simpler implementation is an advantage.